File: backend/app/api/routers/mmm.py
"""Midweek Monster Mash donor-badge router (public, read-only).

  GET /api/mmm/badges        -> the badge-tier catalog (title + point value)
  GET /api/mmm/donors        -> the ranked supporter leaderboard
  GET /api/mmm/donors/{name} -> one supporter's standing (404 if not listed)
"""

from fastapi import APIRouter

from app.api.dependencies.services import MmmServiceDep
from app.schemas.mmm import BadgeTier, DonorSummary

router = APIRouter(prefix="/mmm", tags=["mmm"])


@router.get("/badges", response_model=list[BadgeTier], summary="Badge tier catalog")
def badges(service: MmmServiceDep) -> list[BadgeTier]:
    return service.badges()


@router.get("/donors", response_model=list[DonorSummary], summary="Donor leaderboard")
def donors(service: MmmServiceDep) -> list[DonorSummary]:
    return service.leaderboard()


@router.get("/donors/{name}", response_model=DonorSummary, summary="A single donor by name")
def donor(name: str, service: MmmServiceDep) -> DonorSummary:
    return service.donor(name)

File: frontend/src/api/mmm.ts
import { apiClient } from './client';
import type { MmmBadgeTier, MmmDonor, MmmImportResult } from '@/types';

export const mmmApi = {
  // The badge-tier catalog (lowest → highest).
  badges: async (): Promise<MmmBadgeTier[]> => {
    const { data } = await apiClient.get<MmmBadgeTier[]>('/mmm/badges');
    return data;
  },

  // The ranked supporter leaderboard.
  donors: async (): Promise<MmmDonor[]> => {
    const { data } = await apiClient.get<MmmDonor[]>('/mmm/donors');
    return data;
  },

  // One supporter by name (404 if not listed).
  donor: async (name: string): Promise<MmmDonor> => {
    const { data } = await apiClient.get<MmmDonor>(`/mmm/donors/${encodeURIComponent(name)}`);
    return data;
  },

  // Admin only: replace donor data from an uploaded CSV.
  importDonors: async (file: File): Promise<MmmImportResult> => {
    const form = new FormData();
    form.append('file', file);
    const { data } = await apiClient.post<MmmImportResult>('/mmm/admin/donors', form);
    return data;
  },
};
